Refactor: Improve setup_clean_chain semantics

fjahr opened the pull request:
The semantics of `setup_clean_chain` are off in my opinion. If it is set to true, there is actually no real chain setup, we only start with the Genesis block. That is not what I would expect 'clean' to mean. The 200 blocks chain that we set up if it is set to false could also be called 'clean' since there are no transactions or reorgs.

I think this is also the reason that other people have been confused. The documentation for `test_shell` is not correct and the default has been re-declared in several tests. This PR fixes the documentation and removes settings that don't change the default. Additionally I suggest renaming `setup_clean_chain` to `skip_chain_setup`.

MarcoFalke commented:
The first commit can also be a scripted diff with the sed `d` action?

MarcoFalke commented:
ACK 6e1a6ad68a9ebf3bc114cc40c96c5075937a93d4

fjahr commented:
> The first commit can also be a scripted diff with the sed `d` action?

Yepp, done.

jonatack commented on test/functional/example_test.py:
```diff
@@ -78,7 +78,10 @@ def set_test_params(self):
 
         This method must be overridden and num_nodes must be explicitly set."""
-        self.setup_clean_chain = True
+        # By default the test pre-mines a chain of 200 blocks automatically.
```
The test pre-mines, or it loads the cached data directories?

fjahr replied:
Yeah, it loads a pre-mined chain. Given that mostly absolute beginners are going to look at this I was trying to make it easier to understand but in this case was giving the wrong impression. Updated.

jonatack reviewed (commented):
My issue with this setting is not how it's named, but the lack of clarity on whether the default setting (e.g. `setup_clean_chain=False`) is the efficient one, and if it is not, why is the inefficient setting the default and what the real-world tradeoffs and differences are -- and to clarify all this better in the README and the example test. I don't see this renaming making that clearer but would love for the docs to be clearer.

MarcoFalke commented:
@jonatack If you need the blocks in the chain, then loading them from the cache is efficient. If you don't need them, then you may skip the cache, but it wouldn't hurt if you forget to skip the cache.

Setting the default the other way round would risk that tests that need blocks generate them instead of loading them from the cache.

jonatack commented:
Thanks @MarcoFalke, so IIUC the default is for safety. @fjahr from my personal perspective, adding helpful discussion in the docs about the tradeoffs and reasons would be very helpful (I'm less enthusiastic about renaming).

MarcoFalke commented:
I like the rename, but naming is always subjective to some extend

jonatack commented:
Maybe I'm thoroughly confused, but isn't the new name inaccurate? :) ISTM it ought to be strictly better to justify a change. For the docs, there is also this recent conversation https://github.com/bitcoin/bitcoin/pull/18638#discussion_r429623460 about when to use this. It's very late here but will revisit this at a better hour and try to come up with a clear addition to the docs that everyone agrees with, unless @fjahr you're keen to do it (which would be great).

fjahr commented:
> Maybe I'm thoroughly confused, but isn't the new name inaccurate? :) ISTM it ought to be strictly better to justify a change. For the docs, there is also this recent conversation [#18638 (comment)](https://github.com/bitcoin/bitcoin/pull/18638#discussion_r429623460) about when to use this. It's very late here but will revisit this at a better hour and try to come up with a clear addition to the docs that everyone agrees with, unless @fjahr you're keen to do it (which would be great).

It is a different abstraction layer. The person writing a test using `BitcoinTestFramework` should not have to learn every little detail of how the framework works internally. That's why it is a Framework. I mentioned why I think the name change is better: because the current name is very confusing. I think most people do not understand 'clean' to be what it is supposed to mean here. I am trying to use a name that is easy to understand and appropriate for the abstraction layer and I think setup is a neutral term that does not imply mining or loading from cache.

Feel free to open an additional PR for more docs updates or give me a hint where you think I could extend this here and I will add it.

fjahr commented:
Added some more small adjustments to the docs based on feedback from @jonatack to make sure users understand performance implications.

jonatack commented on test/functional/example_test.py:
```diff
@@ -78,7 +78,10 @@ def set_test_params(self):
         This method must be overridden and num_nodes must be explicitly set."""
-        self.setup_clean_chain = True
+        # By default the test loads a pre-mined chain of 200 blocks from cache.
+        # Set skip_chain_setup to True to skip this and start from the Genesis
```
"True to not do something" seems like a naming regression. True to set up a clean chain is clearer.

jonatack reviewed (commented):
I think this renaming is unwise and am NACK on it. Why:

- Semi-subjective: The new name is more confusing. "Set to true to not do something" is not an improvement in clarity.

- Objective: The change has consequences that are strictly worse. Changing all these files is noise and churn, messing with git history and erasing all usability of git blame for these settings across the functional test suite. These aren't copyright headers but test settings that are important enough that the convention has been to incude them explicitly in most every file. The only wise reason to change these lines is to change the setting.

Concept ACK on improving the documentation. Frameworks increase, not decrease, the need for good documentation and clear explanation of when and why to use settings. Links to source code or terms to grep should be added. Reviewers and contributors aren't beginners to hide details from; they need to thoroughly understand what is happening under the hood. That is why this documentation exists.

MarcoFalke commented on test/functional/example_test.py:
```diff
@@ -78,7 +78,10 @@ def set_test_params(self):
+        # By default the test loads a pre-mined chain of 200 blocks from cache.
+        # Set skip_chain_setup to True to skip this and start from the Genesis
+        # block.
+        self.skip_chain_setup = True
```
```suggestion
        self.use_cached_chain = False
```

The double negative is a good point. What about this name?

fjahr replied:
Sounds good to me, happy to change it to that if other reviewers prefer it. :)

MarcoFalke replied:
> Sounds good to me, happy to change it to that if other reviewers prefer it. :)


It would make me feel less guilty of picking the "clean chain" name in fac93497986b5f74716383bf26c78406253c625a, so I hope that other reviewers take that into consideration :sweat_smile:

fjahr commented:
> * Semi-subjective: The new name is more confusing. "Set to true to not do something" is not an improvement in clarity.

I don't understand this argument. The naming itself is misleading. So you would like it to be clearly misleading rather than not so clearly non-misleading? Anyway, I think @MarcoFalke 's suggested renaming would solve this. 

> * Objective: The change has consequences that are strictly worse. Changing all these files is noise and churn, messing with git history and erasing all usability of git blame for these settings across the functional test suite. These aren't copyright headers but test settings that are important enough that the convention has been to include them explicitly in almost every file. The only wise reason to change these lines is to change the setting.

If we would follow this rule blindly we would never do any refactorings. However, we do them sometimes if there is an added benefit. The benefit here is clarity of what this setting is doing. There is no doubt it's confusing as I have laid out before. And in this case, it is also less critical because this line never stands on its own. It is setup code that enables other stuff to work but is never a standalone thing that is relevant without the other code it makes work. The other will be the lines that developers will want to `git blame`, not the setup code. At least that is my impression of the code I have seen in the files I am changing. Feel free to give counterexamples if I am wrong.

MarcoFalke commented:
btw, `git blame` on recent versions of git can ignore specific commits (like refactoring scripted diffs)

jonatack commented:
> > * Objective: The change has consequences that are strictly worse. Changing all these files is noise and churn, messing with git history and erasing all usability of git blame for these settings across the functional test suite. These aren't copyright headers but test settings that are important enough that the convention has been to include them explicitly in almost every file. The only wise reason to change these lines is to change the setting.
> 
> If we would follow this rule blindly we would never do any refactorings.

I am not suggesting to follow this blindly, but to follow it in this case for the reasons provided.

jonatack commented:
I suggest improving the documentation that can be pointed to in the future regarding the use of this setting.

fjahr commented:
rebased

fjahr commented:
> I suggest improving the documentation that can be pointed to in the future regarding the use of this setting.

Doesn't 5d94c2ebf70a3bc4634ff97da18a19040fd95cbb do that?

fjahr committed:
scripted-diff: Remove setup_clean_chain if default is not changed

-BEGIN VERIFY SCRIPT-
git grep -l "self.setup_clean_chain = False" test/functional/*.py | xargs sed -i "/self.setup_clean_chain = False/d";
-END VERIFY SCRIPT-

fjahr committed:
doc: Clean up use_cached_chain renaming in docs

Also clarify behavior better.

fjahr committed:
scripted-diff: Rename setup_clean_chain to use_cached_chain

Also switched the effect of the bool value, i.e. setup_clean_chain=True is now use_cached_chain=False.

-BEGIN VERIFY SCRIPT-
git grep -l "self.setup_clean_chain = True" test/functional | xargs sed -i "s/self.setup_clean_chain = True/self.use_cached_chain = False/g";

git grep -l "self.setup_clean_chain = False" test/functional/test_framework | xargs sed -i "s/self.setup_clean_chain = False/self.use_cached_chain = True/g";

git grep -l "not self.setup_clean_chain" test/functional/test_framework | xargs sed -i "s/not self.setup_clean_chain/self.use_cached_chain/g";

git grep -l "self.setup_clean_chain" test/functional/test_framework | xargs sed -i "s/self.setup_clean_chain/not self.use_cached_chain/g";

git grep -l "setup_clean_chain" test/functional/test_framework | xargs sed -i "s/setup_clean_chain/use_cached_chain/g";
-END VERIFY SCRIPT-

fjahr commented:
Giving this one last effort with the rename to `use_chached_chain` as suggested by @MarcoFalke . Rebased on master to fix silent merge conflicts.

fjahr commented:
Rebased. @jonatack could you check if the latest changes make you reconsider your NACK? Incorrect use of the `setup_clean_chain` config still happens regularly (and slips though reviews) as in this commit that was just recently merged: https://github.com/bitcoin/bitcoin/commit/3bd67ba5a4ef3c20ef1f873b63c9f53a6c8608b6.

jonatack commented:
Sure, sorry about that. Re-reviewing soon.

gzhao408 commented on test/functional/example_test.py:
```diff
@@ -78,7 +78,10 @@ def set_test_params(self):
+        # By default the test loads a pre-mined chain of 200 blocks from cache.
+        # Set use_chached_chain to False to skip this and start from the Genesis
+        # block.
+        self.use_cached_chain = False
```
Noob confession: when I first started tinkering with example_test.py, I struggled a lot because I didn’t quite understand what a “clean chain” meant and wasn't aware that coinbases needed to mature (I was trying to create transactions with `sendtoaddress` and getting "insufficient funds"). Maybe comments/docs here or elsewhere could help?

fjahr replied:
Hm, I get your point but I think explaining coinbase maturity, a very basic part of the protocol, here in the example test or in the functional test README, feels out of place to me.

glozow reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/19168/commits/ee74d2398e4753124b442701d61d8a8febec5ab1

For me personally, I think this is clearer: if I want to start the chain from scratch (which is not the default), I set the option to False. 

I have a question about why `use_cached_chain` is the default instead of the other way around? It seems appropriate to me that, when we’re trying to test bitcoind behavior, starting with an existing chain and wallets funded represents a typical situation. However, it seems like starting from scratch is more common than using the cached chain in functional tests? I think using the cached chain is more convenient, but are there other reasons to use it?

fjahr commented:
> I have a question about why `use_cached_chain` is the default instead of the other way around? It seems appropriate to me that, when we’re trying to test bitcoind behavior, starting with an existing chain and wallets funded represents a typical situation. However, it seems like starting from scratch is more common than using the cached chain in functional tests? I think using the cached chain is more convenient, but are there other reasons to use it?

I think @MarcoFalke mostly answered that question here: https://github.com/bitcoin/bitcoin/pull/19168#issuecomment-639070665 The downside of not using the cached chain when you could is bigger than using it when you don't need it.

jonatack reviewed (commented):
Concept ACK. I'd suggest:

- merging the two manual doc edit commits into one after the scripted diffs

- possibly automating the misspelled manual renaming by adding it to the scripted diff

- possibly dropping the commit that removes the default setting from tests, the rationale being that it is helpful and saves time when running the test with the opposite setting, when it is present.

fjahr commented:
Had to do another rebase, no code changes `git range-diff e9b30126545d6ddd8772363e4079d1e4908ad117..55e5bc178cf9e562a36d204c79a01a0d7db1a066 bab4cce1b0eedc1a51692aaf83ba54dd0a9d17e6..708bfc31388cdb7fcf8b3348f012e6753116bfbc`. Would you mind taking another look @MarcoFalke @gzhao408 ?

fjahr commented:
Ain't no party like a rebase party, 'cause a rebase party don't stop...

fjahr commented:
Rebased

fjahr commented:
herwaardeer (rebased in Afrikaans)

fjahr commented:
Closing this since the renaming seems to be too controversial/is not interesting enough to most reviewers. I opened a new PR without the renaming and some improved doc cleanups in #21042.
